src/PdfViewer/index.tsx

```diff
@@ -0,0 +1,97 @@
+import 'react-pdf/dist/esm/Page/AnnotationLayer.css';
+import 'react-pdf/dist/esm/Page/TextLayer.css';
+
+import { FileEntry } from "@tauri-apps/api/fs";
+import { useCallback, useEffect, useLayoutEffect, useRef, useState } from "react";
+import { Document, Page } from "react-pdf";
+
+import { readFile } from "../filesystem";
+import { PdfViewerAPI } from '../types/PdfViewerAPI';
+
+export interface PdfViewerProps {
+  file: FileEntry;
+  pdfViewerRef: React.MutableRefObject<PdfViewerAPI | null>;
+}
+
+export const PdfViewer = ({ file, pdfViewerRef }: PdfViewerProps) => {
+  const [numPages, setNumPages] = useState<number | null>(null);
+
+  const [fileContent, setFileContent] = useState<{ data: Uint8Array } | null>(null);
+  const [isFileLoading, setIsFileLoading] = useState(true);
+  const [error, setError] = useState<string | null>(null);
+
+  const containerRef = useRef<HTMLDivElement>(null);
+  const [pdfViewerWidth, setPdfViewerWidth] = useState<number>();
+
+  const updateTimeoutRef = useRef<NodeJS.Timeout>();
+
+
+  const updateWidth = useCallback(() => {
+    clearTimeout(updateTimeoutRef.current);
+    updateTimeoutRef.current = setTimeout(() => {
+      setPdfViewerWidth(containerRef.current?.getBoundingClientRect().width);
+    }, 200);
+  }, [updateTimeoutRef]);
+
+  // Update viewer's width on mount
+  useLayoutEffect(updateWidth, [updateWidth]);
+
+  // Update viewer's width on window resize
+  useEffect(() => {
+    window.addEventListener('resize', updateWidth);
+
+    return () => {
+      window.removeEventListener('resize', updateWidth);
+    };
+  }, [updateWidth]);
+
+  // Update viewer's width on panel resize
+  useEffect(() => {
+    pdfViewerRef.current = { updateWidth };
+
+    return () => {
+      pdfViewerRef.current = null;
+    }
+  }, [pdfViewerRef, updateWidth]);
+
+  useEffect(() => {
+    setNumPages(null);
+    setIsFileLoading(true);
+    (async () => {
+      try {
+        const content = await readFile(file);
+        setFileContent({ data: content });
+        setIsFileLoading(false);
+      } catch (err) {
+        if (err instanceof Error) {
+          setError(err.message);
+        } else {
+          console.error(err);
+          setError('Unknown error');
+        }
+      }
+    })()
+  }, [file]);
+
+  const onDocumentLoadSuccess = ({ numPages }: { numPages: number }) => {
+    setNumPages(numPages);
+  }
+
+  if (error) return <div>An error occured while opening the selected file: {error}</div>;
+
+  if (isFileLoading) return <div><strong>Loading</strong></div>;
+
+  return (
+    <div className="pdf-viewer flex flex-col h-full" ref={containerRef}>
+      <Document
+        file={fileContent}
+        onLoadSuccess={onDocumentLoadSuccess}
+        className="flex-1 overflow-scroll"
+      >
+        {numPages && Array.from(new Array(numPages), (_, index) => (
+          <Page key={`page_${index + 1}`} pageNumber={index + 1} width={pdfViewerWidth} />
+        ))}
```

Comment:
Why do we need this part?

Reply:
This comes from the documentation: https://github.com/wojtekmaj/react-pdf/wiki/Recipes#display-all-pages

Basically, `Document` is just a context provider and displays nothing. Then you have to add the `Page`s. You can either have a the current page as state in your component and only display this page, along with buttons to change the current page, or you display all the pages at once (which is what we want here). Unfortunately, javascript has no clean way to create an array [1, ..., numPages], which is why we have to do the `Array.from(new Array(numPages), (_, index) => (...` "hack".